packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/MrwSerializationTypeProvider.cs

```diff
@@ -294,6 +999,521 @@ private MethodBodyStatement GetPropertyInitializers(IReadOnlyList<ParameterProvi
+
+                // Generate the serialization statements for the property
+                var writePropertySerializationStatements = new MethodBodyStatement[]
+                {
+                    utf8JsonWriter.WritePropertyName(PropertySerializations[i].VariableName),
+                    CreateSerializationStatement(utf8JsonWriter, PropertySerializations[i].Type, propertyMember, PropertySerializations[i].SerializationFormat, options)
+                };
+
+                // Wrap the serialization statement in a check for whether the property is defined
+                var wrapInIsDefinedStatement = WrapInIsDefined(utf8JsonWriter, PropertySerializations[i], propertyMember, writePropertySerializationStatements);
+                propertyStatements[i] = inputProperty.IsReadOnly ? WrapInCheckNotWireIfStatement(options.Format, wrapInIsDefinedStatement) : wrapInIsDefinedStatement;
+            }
+
+            return propertyStatements;
+        }
+
+        private MethodBodyStatement CheckPropertyIsInitialized(
+            Utf8JsonWriterSnippet utf8JsonWriter,
+            PropertySerializationProvider propertySerializationProvider,
+            MemberExpression propertyMemberExpression,
+            MethodBodyStatement writePropertySerializationStatements)
+        {
+            var propertyType = propertySerializationProvider.Type;
+            BoolSnippet propertyIsInitialized;
+            if (propertyType.IsCollection && !propertyType.IsReadOnlyMemory && propertySerializationProvider.IsRequired)
+            {
+                propertyIsInitialized = And(NotEqual(propertyMemberExpression, Null),
+                    Optional.IsCollectionDefined(new StringSnippet(propertyMemberExpression)));
+            }
+            else
+            {
+                propertyIsInitialized = NotEqual(propertyMemberExpression, Null);
+            }
+
+            return new IfElseStatement(
+                propertyIsInitialized,
+                writePropertySerializationStatements,
+                utf8JsonWriter.WriteNull(propertySerializationProvider.SerializedName));
+        }
+
+        /// <summary>
+        /// Wraps the serialization statement in a condition check to ensure only initialized and required properties are serialized.
+        /// </summary>
+        /// <param name="utf8JsonWriter">The JSON writer to use for serialization.</param>
+        /// <param name="propertySerializationProvider">Provides serialization settings for the property.</param>
+        /// <param name="propertyMemberExpression">The expression representing the property to serialize.</param>
+        /// <param name="writePropertySerializationStatement">The serialization statement to conditionally execute.</param>
+        /// <returns>A method body statement that includes condition checks before serialization.</returns>
+        private MethodBodyStatement WrapInIsDefined(
```

Comment:
This feels like a duplicate of https://github.com/microsoft/typespec/blob/main/packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/OptionalProvider.cs#L167 is there any reason we don't use this directly instead?